File: 2021/07/31/SingleNumber.java
/**
 * https://leetcode.com/problems/single-number/
 * @author Andrew Jarombek
 * @since 7/31/2021
 */

import java.util.HashSet;

public class SingleNumber {
    public static int singleNumber(int[] nums) {
        HashSet<Integer> appearsOnce = new HashSet<>();
        HashSet<Integer> appearsTwice = new HashSet<>();

        for (int num: nums) {
            if (appearsOnce.contains(num)) {
                appearsOnce.remove(num);
                appearsTwice.add(num);
            } else {
                appearsOnce.add(num);
            }
        }

        for (int result: appearsOnce) {
            return result;
        }

        return 0;
    }

    public static void main(String[] args) {
        assert singleNumber(new int[] {2, 2, 1}) == 1;
    }
}

File: 2022/02/16/c/singleNumber.c
// Program: https://leetcode.com/problems/single-number/
// Author: Andrew Jarombek
// Date: 2/16/2022

#include <stdlib.h>
#include <assert.h>

int singleNumber(int* nums, int numsSize){
    int value = 0;

    for (int i = 0; i < numsSize; i++) {
        value ^= nums[i];
    }

    return value;
}

int main() {
    int* array = (int*) malloc(sizeof(int) * 5);
    array[0] = 2;
    array[1] = 2;
    array[2] = 1;

    assert(singleNumber(array, 3) == 1);

    free(array);
}
